File: fall_simulation.py
"""
Ejercicio 5: Simulación de caída libre para múltiples objetos
Implementación paralela usando concurrent.futures para calcular tiempos de caída independientes
"""

import time
import random
import math
from concurrent.futures import ProcessPoolExecutor, ThreadPoolExecutor
from typing import List, Tuple, NamedTuple
import numpy as np


class FallingObject(NamedTuple):
    """Representa un objeto en caída libre"""
    id: int
    height: float  # Altura inicial en metros
    mass: float    # Masa en kg
    drag_coefficient: float  # Coeficiente de resistencia del aire


class FallResult(NamedTuple):
    """Resultado del cálculo de caída libre"""
    object_id: int
    height: float
    fall_time: float
    final_velocity: float
    processing_time: float


def calculate_fall_time(object_data: FallingObject) -> FallResult:
    """
    Calcula el tiempo de caída libre para un objeto
    
    Args:
        object_data: Datos del objeto en caída libre
    
    Returns:
        Resultado del cálculo
    """
    start_time = time.time()
    
    # Constantes físicas
    g = 9.81  # Aceleración gravitacional (m/s²)
    air_density = 1.225  # Densidad del aire (kg/m³)
    
    height = object_data.height
    mass = object_data.mass
    drag_coeff = object_data.drag_coefficient
    
    # Cálculo simplificado del tiempo de caída considerando resistencia del aire
    # Para objetos esféricos, área proyectada aproximada
    radius = (mass / (4/3 * math.pi * 1000))**(1/3)  # Asumiendo densidad del agua
    area = math.pi * radius**2
    
    # Fuerza de resistencia del aire
    drag_force = 0.5 * air_density * drag_coeff * area
    
    # Ecuación simplificada para tiempo de caída con resistencia del aire
    # Usando aproximación numérica simple
    dt = 0.001  # Paso de tiempo pequeño
    t = 0
    v = 0
    y = height
    
    while y > 0:
        # Fuerza neta (gravedad - resistencia del aire)
        net_force = mass * g - drag_force * v**2
        acceleration = net_force / mass
        
        # Actualizar velocidad y posición
        v += acceleration * dt
        y -= v * dt
        t += dt
        
        # Evitar bucles infinitos
        if t > 100:  # Máximo 100 segundos
            break
    
    # Velocidad final aproximada
    final_velocity = v
    
    processing_time = time.time() - start_time
    
    return FallResult(
        object_id=object_data.id,
        height=height,
        fall_time=t,
        final_velocity=final_velocity,
        processing_time=processing_time
    )


def simulate_fall_sequential(objects: List[FallingObject]) -> List[FallResult]:
    """
    Versión secuencial de simulación de caída libre
    
    Args:
        objects: Lista de objetos a simular
    
    Returns:
        Lista de resultados
    """
    results = []
    
    for obj in objects:
        result = calculate_fall_time(obj)
        results.append(result)
    
    return results


def simulate_fall_parallel(objects: List[FallingObject], max_workers: int = None) -> List[FallResult]:
    """
    Versión paralela de simulación de caída libre usando ProcessPoolExecutor
    
    Args:
        objects: Lista de objetos a simular
        max_workers: Número máximo de workers
    
    Returns:
        Lista de resultados
    """
    with ProcessPoolExecutor(max_workers=max_workers) as executor:
        results = list(executor.map(calculate_fall_time, objects))
    
    return results


def create_sample_objects(num_objects: int = 10) -> List[FallingObject]:
    """
    Crea objetos de muestra para testing
    
    Args:
        num_objects: Número de objetos a crear
    
    Returns:
        Lista de objetos en caída libre
    """
    objects = []
    
    for i in range(num_objects):
        # Generar parámetros aleatorios realistas
        height = random.uniform(10, 200)  # Altura entre 10 y 200 metros
        mass = random.uniform(0.1, 50)    # Masa entre 0.1 y 50 kg
        drag_coeff = random.uniform(0.1, 1.5)  # Coeficiente de resistencia
        
        obj = FallingObject(
            id=i + 1,
            height=height,
            mass=mass,
            drag_coefficient=drag_coeff
        )
        objects.append(obj)
    
    return objects


def benchmark_fall_simulation():
    """
    Ejecuta benchmark comparando versión secuencial vs paralela
    """
    print("=== EJERCICIO 5: SIMULACIÓN DE CAÍDA LIBRE ===\n")
    
    # Crear objetos de muestra
    print("Creando objetos de muestra...")
    objects = create_sample_objects(12)
    print(f"Se crearon {len(objects)} objetos para simular\n")
    
    # Mostrar información de los objetos
    print("Objetos creados:")
    for obj in objects:
        print(f"  Objeto {obj.id}: altura={obj.height:.1f}m, masa={obj.mass:.2f}kg, drag={obj.drag_coefficient:.2f}")
    print()
    
    # Benchmark versión secuencial
    print("Ejecutando versión secuencial...")
    start_time = time.time()
    sequential_results = simulate_fall_sequential(objects)
    sequential_time = time.time() - start_time
    
    print(f"Tiempo secuencial: {sequential_time:.4f} segundos")
    print(f"Tiempo promedio por objeto: {sequential_time/len(objects):.4f} segundos")
    
    # Calcular estadísticas secuenciales
    avg_fall_time_seq = sum(result.fall_time for result in sequential_results) / len(sequential_results)
    avg_velocity_seq = sum(result.final_velocity for result in sequential_results) / len(sequential_results)
    
    print(f"Tiempo promedio de caída (secuencial): {avg_fall_time_seq:.2f}s")
    print(f"Velocidad promedio final (secuencial): {avg_velocity_seq:.2f}m/s")
    
    # Benchmark versión paralela
    print("\nEjecutando versión paralela...")
    start_time = time.time()
    parallel_results = simulate_fall_parallel(objects)
    parallel_time = time.time() - start_time
    
    print(f"Tiempo paralelo: {parallel_time:.4f} segundos")
    print(f"Tiempo promedio por objeto: {parallel_time/len(objects):.4f} segundos")
    
    # Calcular estadísticas paralelas
    avg_fall_time_par = sum(result.fall_time for result in parallel_results) / len(parallel_results)
    avg_velocity_par = sum(result.final_velocity for result in parallel_results) / len(parallel_results)
    
    print(f"Tiempo promedio de caída (paralelo): {avg_fall_time_par:.2f}s")
    print(f"Velocidad promedio final (paralelo): {avg_velocity_par:.2f}m/s")
    
    # Verificar que los resultados sean iguales
    print(f"\nVerificación de resultados:")
    time_diff = abs(avg_fall_time_seq - avg_fall_time_par)
    velocity_diff = abs(avg_velocity_seq - avg_velocity_par)
    print(f"Diferencia en tiempo promedio: {time_diff:.6f}s")
    print(f"Diferencia en velocidad promedio: {velocity_diff:.6f}m/s")
    
    # Calcular speedup
    speedup = sequential_time / parallel_time if parallel_time > 0 else 0
    print(f"\nSpeedup: {speedup:.2f}x")
    
    # Mostrar resultados detallados
    print("\nResultados detallados:")
    for result in parallel_results:
        print(f"  Objeto {result.object_id}: altura={result.height:.1f}m, "
              f"tiempo_caída={result.fall_time:.2f}s, "
              f"velocidad_final={result.final_velocity:.2f}m/s, "
              f"tiempo_procesamiento={result.processing_time:.4f}s")
    
    print(f"\nObjetos simulados exitosamente: {len(parallel_results)}")
    print("\nNota: Los cálculos incluyen resistencia del aire para mayor realismo\n")


if __name__ == "__main__":
    benchmark_fall_simulation()

File: fall_simulation.cpp
/*
 * Ejercicio 5: Simulación de caída libre para múltiples objetos
 * Implementación paralela usando OpenMP para calcular tiempos de caída independientes
 */

#include <iostream>
#include <vector>
#include <chrono>
#include <random>
#include <cmath>
#include <omp.h>
#include <iomanip>

using namespace std;

// Estructura para representar un objeto en caída libre
struct FallingObject {
    int id;
    double height;           // Altura inicial en metros
    double mass;             // Masa en kg
    double drag_coefficient; // Coeficiente de resistencia del aire
};

// Estructura para almacenar el resultado del cálculo de caída libre
struct FallResult {
    int object_id;
    double height;
    double fall_time;
    double final_velocity;
    double processing_time;
};

// Función para calcular el tiempo de caída libre para un objeto
FallResult calculateFallTime(const FallingObject& obj) {
    auto start_time = chrono::high_resolution_clock::now();
    
    // Constantes físicas
    const double g = 9.81;        // Aceleración gravitacional (m/s²)
    const double air_density = 1.225; // Densidad del aire (kg/m³)
    
    double height = obj.height;
    double mass = obj.mass;
    double drag_coeff = obj.drag_coefficient;
    
    // Cálculo simplificado del tiempo de caída considerando resistencia del aire
    // Para objetos esféricos, área proyectada aproximada
    double radius = pow(mass / (4.0/3.0 * M_PI * 1000), 1.0/3.0); // Asumiendo densidad del agua
    double area = M_PI * radius * radius;
    
    // Fuerza de resistencia del aire
    double drag_force = 0.5 * air_density * drag_coeff * area;
    
    // Ecuación simplificada para tiempo de caída con resistencia del aire
    // Usando aproximación numérica simple
    double dt = 0.001; // Paso de tiempo pequeño
    double t = 0;
    double v = 0;
    double y = height;
    
    while (y > 0) {
        // Fuerza neta (gravedad - resistencia del aire)
        double net_force = mass * g - drag_force * v * v;
        double acceleration = net_force / mass;
        
        // Actualizar velocidad y posición
        v += acceleration * dt;
        y -= v * dt;
        t += dt;
        
        // Evitar bucles infinitos
        if (t > 100) { // Máximo 100 segundos
            break;
        }
    }
    
    // Velocidad final aproximada
    double final_velocity = v;
    
    auto end_time = chrono::high_resolution_clock::now();
    auto duration = chrono::duration_cast<chrono::microseconds>(end_time - start_time);
    double processing_time = duration.count() / 1000000.0; // Convertir a segundos
    
    return FallResult{
        obj.id,
        height,
        t,
        final_velocity,
        processing_time
    };
}

// Versión secuencial de simulación de caída libre
vector<FallResult> simulateFallSequential(const vector<FallingObject>& objects) {
    vector<FallResult> results;
    
    for (const auto& obj : objects) {
        results.push_back(calculateFallTime(obj));
    }
    
    return results;
}

// Versión paralela de simulación de caída libre usando OpenMP
vector<FallResult> simulateFallParallel(const vector<FallingObject>& objects) {
    vector<FallResult> results(objects.size());
    
    #pragma omp parallel for
    for (int i = 0; i < static_cast<int>(objects.size()); ++i) {
        results[i] = calculateFallTime(objects[i]);
    }
    
    return results;
}

// Función para crear objetos de muestra
vector<FallingObject> createSampleObjects(int num_objects = 12) {
    vector<FallingObject> objects;
    random_device rd;
    mt19937 gen(rd());
    uniform_real_distribution<double> height_dis(10.0, 200.0);
    uniform_real_distribution<double> mass_dis(0.1, 50.0);
    uniform_real_distribution<double> drag_dis(0.1, 1.5);
    
    for (int i = 0; i < num_objects; ++i) {
        FallingObject obj;
        obj.id = i + 1;
        obj.height = height_dis(gen);        // Altura entre 10 y 200 metros
        obj.mass = mass_dis(gen);            // Masa entre 0.1 y 50 kg
        obj.drag_coefficient = drag_dis(gen); // Coeficiente de resistencia
        
        objects.push_back(obj);
    }
    
    return objects;
}

// Función principal de benchmark
void benchmarkFallSimulation() {
    cout << "=== EJERCICIO 5: SIMULACIÓN DE CAÍDA LIBRE (C++ con OpenMP) ===" << endl << endl;
    
    // Crear objetos de muestra
    cout << "Creando objetos de muestra..." << endl;
    vector<FallingObject> objects = createSampleObjects(12);
    cout << "Se crearon " << objects.size() << " objetos para simular" << endl << endl;
    
    // Mostrar información de los objetos
    cout << "Objetos creados:" << endl;
    for (const auto& obj : objects) {
        cout << "  Objeto " << obj.id << ": altura=" << fixed << setprecision(1) << obj.height 
             << "m, masa=" << setprecision(2) << obj.mass 
             << "kg, drag=" << obj.drag_coefficient << endl;
    }
    cout << endl;
    
    // Benchmark versión secuencial
    cout << "Ejecutando versión secuencial..." << endl;
    auto start_time = chrono::high_resolution_clock::now();
    vector<FallResult> sequential_results = simulateFallSequential(objects);
    auto end_time = chrono::high_resolution_clock::now();
    auto sequential_duration = chrono::duration_cast<chrono::microseconds>(end_time - start_time);
    double sequential_time = sequential_duration.count() / 1000000.0;
    
    cout << "Tiempo secuencial: " << fixed << setprecision(4) << sequential_time << " segundos" << endl;
    cout << "Tiempo promedio por objeto: " << sequential_time / objects.size() << " segundos" << endl;
    
    // Calcular estadísticas secuenciales
    double avg_fall_time_seq = 0.0;
    double avg_velocity_seq = 0.0;
    for (const auto& result : sequential_results) {
        avg_fall_time_seq += result.fall_time;
        avg_velocity_seq += result.final_velocity;
    }
    avg_fall_time_seq /= sequential_results.size();
    avg_velocity_seq /= sequential_results.size();
    
    cout << "Tiempo promedio de caída (secuencial): " << fixed << setprecision(2) << avg_fall_time_seq << "s" << endl;
    cout << "Velocidad promedio final (secuencial): " << avg_velocity_seq << "m/s" << endl;
    
    // Benchmark versión paralela
    cout << endl << "Ejecutando versión paralela..." << endl;
    start_time = chrono::high_resolution_clock::now();
    vector<FallResult> parallel_results = simulateFallParallel(objects);
    end_time = chrono::high_resolution_clock::now();
    auto parallel_duration = chrono::duration_cast<chrono::microseconds>(end_time - start_time);
    double parallel_time = parallel_duration.count() / 1000000.0;
    
    cout << "Tiempo paralelo: " << parallel_time << " segundos" << endl;
    cout << "Tiempo promedio por objeto: " << parallel_time / objects.size() << " segundos" << endl;
    
    // Calcular estadísticas paralelas
    double avg_fall_time_par = 0.0;
    double avg_velocity_par = 0.0;
    for (const auto& result : parallel_results) {
        avg_fall_time_par += result.fall_time;
        avg_velocity_par += result.final_velocity;
    }
    avg_fall_time_par /= parallel_results.size();
    avg_velocity_par /= parallel_results.size();
    
    cout << "Tiempo promedio de caída (paralelo): " << avg_fall_time_par << "s" << endl;
    cout << "Velocidad promedio final (paralelo): " << avg_velocity_par << "m/s" << endl;
    
    // Verificar que los resultados sean iguales
    cout << endl << "Verificación de resultados:" << endl;
    double time_diff = abs(avg_fall_time_seq - avg_fall_time_par);
    double velocity_diff = abs(avg_velocity_seq - avg_velocity_par);
    cout << "Diferencia en tiempo promedio: " << scientific << time_diff << "s" << endl;
    cout << "Diferencia en velocidad promedio: " << velocity_diff << "m/s" << endl;
    
    // Calcular speedup
    double speedup = (parallel_time > 0) ? sequential_time / parallel_time : 0;
    cout << endl << "Speedup: " << fixed << setprecision(2) << speedup << "x" << endl;
    
    int num_threads = omp_get_max_threads();
    double efficiency = (speedup / num_threads) * 100;
    cout << "Eficiencia: " << efficiency << "%" << endl;
    cout << "Número de threads utilizados: " << num_threads << endl;
    
    // Mostrar resultados detallados
    cout << endl << "Resultados detallados:" << endl;
    for (const auto& result : parallel_results) {
        cout << "  Objeto " << result.object_id << ": altura=" << fixed << setprecision(1) << result.height 
             << "m, tiempo_caída=" << setprecision(2) << result.fall_time 
             << "s, velocidad_final=" << result.final_velocity 
             << "m/s, tiempo_procesamiento=" << setprecision(4) << result.processing_time << "s" << endl;
    }
    
    cout << endl << "Objetos simulados exitosamente: " << parallel_results.size() << endl;
    cout << endl << "Nota: Los cálculos incluyen resistencia del aire para mayor realismo" << endl << endl;
}

int main() {
    // Configurar OpenMP
    omp_set_num_threads(omp_get_max_threads());
    
    cout << "OpenMP disponible: " << omp_get_max_threads() << " threads" << endl;
    cout << "Compilador OpenMP: " << _OPENMP << endl << endl;
    
    benchmarkFallSimulation();
    
    return 0;
}
